doc/GeneralCppGuidelines.md

```diff
@@ -0,0 +1,293 @@
+
+**Examples**
+
+Unique pointers: They hold sole ownership over the contained object. It is not possible to have two `unique_ptr` managing the same object.
+```C++
+{
+  std::unique_ptr<Foo> myPtr(new Foo);
+  // Smart pointers allow normal pointer semantics:
+  myPtr->doSomething();  // Calls the member function of Foo
+  f(*myPtr);             // Passes a reference to Foo (or value, depending on the function argument)
+} // myPtr goes out of scope and the instance of Foo is deleted automatically
+
+```
+
+Shared pointers: Shared ownership of the object across several instances of the pointers. The object is deleted when the last `shared_ptr` managing it is destroyed or the last managing `shared_ptr` is assigned another object to hold.
+```C++
+{
+  // Let's assume Bar is derived from Foo:
+  std::shared_ptr<Foo> myPtr1 = std::make_shared<Bar>();
+  {
+    // This myPtr2 points to the same instance of Bar:
+    std::shared_ptr<Foo> myPtr2 = myPtr1;
+  } // myPtr2 goes out of scope, but myPtr1 stays in scope, the instance still exists
+} // Here, the instance of Bar is deleted: The last pointer, myPtr1, goes out of scope
+```
+
+## Constness
+Use `const` wherever you can.
+
+If `const` is used correctly, the compiler helps to identify potential problems.
+
+Functions without side-effects are preferable and should be marked as such. Declare member-functions that do not change the state of the object as `const`.
+
+If your function is only "an observer" of an argument, make it explicit. This way you get compiler errors when a function unexpectedly does change the object's state (if the above rule is followed).
+
+Getter functions should return const values or references. _Note_ that when returning a reference and you cannot guarantee the object's validity after the function call, you risk a dangling reference.
+
+**Example**
+```C++
+class Foo {
+public:
+  // Omitting ctor and dtor
+
+  // The getter does not change the state of Foo
+  const Vector3 position() const {
+    return m_position;
+  }
+
+  // This function does not change the state of Foo or the argument
+  const Vector3 offset(const Vector3& otherPosition) const {
+    return m_position - otherPosition;
+  }
+
+  // This set-function does change the state of Foo not of the argument
+  void setPosition(const Vector3& newPosition) {
+    m_position = newPosition;
+  }
+
+  // This function does change the argument but not Foo
+  void addPositionTo(Vector3& otherPosition) const {
+    otherPosition += m_position;
+  }
+private:
+ Vector3 m_position;
+};
+```
+
+## Lifetime
+Always consider what the lifetime of an entity is and restrict it to the shortest period that makes sense.
+
+Sometimes it is not inherently clear what the lifetime of an object in memory is. While names only exist in the scope in which they are defined, the objects they refer to can exist longer:
+- Objects created with `new` live until they are `delete`d
```

Comment:
I'll rephrase it to "until destroyed using `delete`". I wanted to have the keyword as inline code.